version bump on k8s io client-go and apimachinery

Muzry opened the pull request:
<!-- Thanks for sending a PR! Before submitting:

1. If this is your first PR, check out our contribution guide here https://docs.prylabs.network/docs/contribute/contribution-guidelines
   You will then need to sign our Contributor License Agreement (CLA), which will show up as a comment from a bot in this pull request after you open it. We cannot review code without a signed CLA.
2. Please file an associated tracking issue if this pull request is non-trivial and requires context for our team to understand. All features and most bug fixes should have
   an associated issue with a design discussed and decided upon. Small bug
   fixes and documentation improvements don't need issues.
3. New features and bug fixes must have tests. Documentation may need to
   be updated. If you're unsure what to update, send the PR, and we'll discuss
   in review.
4. Note that PRs updating dependencies and new Go versions are not accepted.
   Please file an issue instead.
5. A changelog entry is required for user facing issues.
-->

**What type of PR is this?**

 Other

**What does this PR do? Why is it needed?**

Updated k8s.io/client-go and k8s.io/apimachinery to v0.34.1.
However, starting from v0.33 (≥ v0.33), FIFO no longer provides the following functions:
- ListKeys()
- GetKeys()
- AddIfNotPresent()

FIFO is an implementation of the k8s.io/client-go/cache/Store interface, but Store itself doesn’t define these functions either. Meanwhile, FIFO is used in:
- beacon-chain/execution/block_cache.go
- beacon-chain/state/stategen/epoch_boundary_state_cache.go
- beacon-chain/cache/sync_committee.go

To preserve the current Prysm implementation and minimize the impact of this change, I copied the FIFO implementation from client-go@v0.32.9.

If there are any concerns with this approach, please let me know.


**Which issues(s) does this PR fix?**

Fixes #15707

**Other notes for review**

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

Muzry committed:
version bump on k8s io client-go and apimachinery

prestonvanloon commented on beacon-chain/cache/common.go:
```diff
@@ -1,11 +1,7 @@
 // trim the FIFO queue to the maxSize.
-func trim(queue *cache.FIFO, maxSize uint64) {
+func trim(queue *FIFO, maxSize uint64) {
 	for s := uint64(len(queue.ListKeys())); s > maxSize; s-- {
```
Since we are copying this library, we should make this better. We don't need to list all keys to get the length of `queue.queue`.

prestonvanloon commented on beacon-chain/cache/k8s_fifo.go:
```diff
@@ -0,0 +1,355 @@
+	"errors"
+	"sync"
+
+	"k8s.io/client-go/tools/cache"
```
If we copied over the things we are importing here, do we need the k8s.io/client-go dependency at all? My brief search indicates this file would be the only import.

prestonvanloon replied:
To be clear, I would recommend only importing things we need. We may not care about the `Queue` interface or the specific errors types.

Muzry replied:
> To be clear, I would recommend only importing things we need. We may not care about the `Queue` interface or the specific errors types.

Sorry for the delayed reply. I agree with your suggestion. I'll update the implementation to only import what we actually need.

prestonvanloon reviewed (commented):
Thanks for the PR. I'm wonder if we can remove these two dependencies entirely.

1) We could copy the FIFO impl and other necessary items. Then add minor improvements.
2) Looks like we only use the yaml tools from k8s.io/apimachinery. Could we just use yaml.v2 or [yaml.v3](https://godoc.org/gopkg.in/yaml.v3)?

The k8s.io/apimachinery dependency was added in #12354, but I don't see a need for this specific implementation of yaml unmarshalling over the dependencies we already had. (cc @james-prysm)

james-prysm commented:
@prestonvanloon yeah let me try removing k8s.io/apimachinery

Muzry committed:
update: fifo minimal implementation

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible

Muzry commented:
@prestonvanloon Sorry for the delayed reply. I've updated the FIFO implementation to be minimal and updated the unit tests to reference FIFO usage. Please take another look. Let me know if there's anything I missed

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible

Muzry committed:
Merge branch 'develop' into fix-k8s-incompatible
